ranger/ranger.go

```diff
@@ -136,8 +140,111 @@ func (r *Ranger) Context() (context.Context, context.CancelFunc) { return r.ctx,
 func (r *Ranger) DB() postgres.DatabaseService                   { return r.db }
 func (r *Ranger) Env() trails.Environment                        { return r.env }
 func (r *Ranger) Metadata() Metadata                             { return r.metadata }
+func (r *Ranger) ReqParser() *req.Parser                         { return r.req }
 func (r *Ranger) SessionStore() session.SessionStorer            { return r.sessions }
 
+// ParseBody reads r.Body into payload and writes to w when this fails.
+//
+// payload must be a pointer to a struct in order receive r.Body's data.
+// Upon success, ParseBody returns true and payload's fields are hydrated with data from r.Body.
+// Upon failure, ParseBody writes to w and returns false.
+// The caller should not write to w when ParseBody fails.
+//
+// ParseBody expects r.Body to be JSON-encoded data.
+// If r.Body is incorrectly formatted JSON, ParseBody writes 400.
+//
+// ParseBody expects the decoded data to pass "validate" struct tags defined on payload's type.
+// If values in the data do not meet validation requirements, ParseBody writes 422.
+//
+// In all cases, if an unhandled error occurs, ParseBody writes 500.
+//
+// If control over response codes is desired, use ranger.ReqParser.ParseBody.
+func (ranger *Ranger) ParseBody(w http.ResponseWriter, r *http.Request, payload any) bool {
+	err := ranger.req.ParseBody(r.Body, payload)
+	if err == nil {
+		return true
+	}
+
+	if errors.Is(err, trails.ErrBadFormat) {
+		if err := ranger.Json(w, r, resp.Code(http.StatusBadRequest)); err != nil {
+			ranger.Error(err.Error(), &logger.LogContext{Error: err, Request: r})
+		}
+
+		return false
+	}
+
+	var verrs *req.ValidationErrors
+	if errors.As(err, &verrs) {
+		args := []resp.Fn{resp.Code(http.StatusUnprocessableEntity)}
+		if !ranger.env.IsProduction() {
+			args = append(args, resp.Data(verrs))
+		}
+
+		if err := ranger.Json(w, r, args...); err != nil {
+			ranger.Error(err.Error(), &logger.LogContext{Error: err, Request: r})
+		}
+
+		return false
+	}
+
+	if err := ranger.Json(w, r, resp.Err(err), resp.Code(http.StatusInternalServerError)); err != nil {
```

Comment:
This is interesting from a logging perspective. Currently, we have a lot of:

```
if err := h.parseRequest(r, req); err != nil {
		h.Json(w, r, Err(err), Code(http.StatusBadRequest))
		return
	}
```

which logs the error and send it to sentry. If any of the scenarios fail (decoding, validating), we get an error. This has often been the preferred behavior since we control the shape of the payload so if an error is triggered in either decoding or validating, these have been high value notifications since I usually failed an edgecase on the frontend.

Since these errors do not get passed back to the calling code, we can't pass the responsibility to the calling code with the current implementation. In archive's case, an error is logged to sentry in every single case of failed parsing/validate flow. Additionally, if decoding succeeded, data is passed to sentry as well. I would be happy to default to logging the error from archive's perspective, but the issue of forwarding data remains as I don't want the entire payload passed to sentry if it has anything that shouldn't show up in logs/sentry. Portal has a handful of scenarios where nothing is logged so maybe those should log or maybe this needs more configurability. It's nice if there is any issue when it shows up as a red log on development as well.

Reply:
I'm not sure I totally follow. If you'd like to log the error in some cases, `ranger.ReqParser.ParseBody` provides direct access to the library so the response can be controlled as well. Is it that you'd like a `ranger` method handling the response writing to also log the error?

For Portal's use case, it looks like the dominant pattern is `h.Json(w, r, Code(http.StatusBadRequest))`. I don't really see a strong throughline on `college-try` for those that diverge from the pattern (i.e., those that do `h.Json(w, r, Err(err), Code(http.StatusBadRequest))` and seems like they should all be refactored out. I haven't seen a need to raise these errors thus far, but we've also had many different clients historically, which haven't been as rock solid as our own ;)

It sounds like, though, configurability is important between the two projects, or, at least, for dev environments. What about a env var that disables or enables logging so there's some flexibility?